concurrent/src/main/java/io/airlift/concurrent/VmRuntimeTicker.java

```diff
@@ -0,0 +1,180 @@
+    {
+        return VM_RUNTIME_TICKER.get();
+    }
+
+    private final Ticker systemTicker;
+    private final long loopSleepMillis;
+    private final long maxSafeVmPauseNanos;
+
+    private final Thread vmRuntimeMonitorThread;
+
+    private final AtomicLong currentValue = new AtomicLong();
+
+    private final AtomicLong totalPauseNanos = new AtomicLong();
+    private final TimeStat pauseTime = new TimeStat();
+
+    // public and volatile to assure the allocation is not optimized away
+    @SuppressWarnings({"WeakerAccess", "PublicField"})
+    public volatile Long sleepObjectAllocation;
+
+    private final AtomicBoolean badSystemNanoTime = new AtomicBoolean();
+
+    public VmRuntimeTicker(Ticker systemTicker)
+    {
+        this(systemTicker, DEFAULT_LOOP_SLEEP_MILLIS, DEFAULT_MAX_SAFE_VM_PAUSE_NANOS);
+    }
+
+    public VmRuntimeTicker(Ticker systemTicker, Duration loopSleep, Duration maxSafeVmPause)
+    {
+        this.systemTicker = requireNonNull(systemTicker, "systemTicker is null");
+
+        this.loopSleepMillis = requireNonNull(loopSleep, "loopSleep is null").toMillis();
+        checkArgument(loopSleepMillis > 0, "loopSleep must be at least 1 millis second");
+
+        this.maxSafeVmPauseNanos = requireNonNull(maxSafeVmPause, "maxSafeVmPause is null").roundTo(NANOSECONDS);
+        checkArgument(maxSafeVmPauseNanos > MILLISECONDS.toNanos(loopSleepMillis) * 2, "maxSafeVmPause must be at least 2 times larger than loopSleep");
+
+        vmRuntimeMonitorThread = new Thread(this::updateLoop, "VM Runtime Monitor");
+        vmRuntimeMonitorThread.setDaemon(true);
+    }
+
+    public void start()
+    {
+        vmRuntimeMonitorThread.start();
+    }
+
+    public void stop()
+    {
+        vmRuntimeMonitorThread.interrupt();
+    }
+
+    @Managed
+    public long getTotalPauseMillis()
+    {
+        return NANOSECONDS.toMillis(totalPauseNanos.get());
+    }
+
+    @Managed
+    public TimeStat getPauseTime()
+    {
+        return pauseTime;
+    }
+
+    @Override
+    public long read()
+    {
+        return currentValue.get();
+    }
+
+    private void updateLoop()
+    {
+        long lastRead = systemTicker.read();
+        while (true) {
+            try {
+                Thread.sleep(loopSleepMillis);
+
+                // Allocate an object to make sure potential allocation stalls are measured.
+                sleepObjectAllocation = lastRead;
+
+                long currentRead = systemTicker.read();
+
+                long sleepDuration = currentRead - lastRead;
+                lastRead = currentRead;
+
+                // check if clock moved backwards
+                if (sleepDuration < 0) {
+                    if (badSystemNanoTime.compareAndSet(false, true)) {
+                        log.warn("Possible kernel or JVM bug: system nano time moved backwards");
+                    }
+                    continue;
+                }
+
+                long sleepDelta = sleepDuration - loopSleepMillis;
```

Comment:
is Thread.sleep mostly guaranteed to return the same or longer than the sleep duration?